Unpin libcxx

**Describe the outcome that is desired.**
Due to pipeline failures we had to pin `libcxx` to `<17` - see #37275 . The desired outcome is to fix the issues and remove this pin.

**Describe any solutions you are considering**
Fix issues around use of `char` in `WideInt.h` and `WideIntImpl.h`

**Additional context**
Examples of failing tests associated with most recent version of `libcxx`
```
11:45:40 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h:183:27: warning: implicit conversion from 'std::numeric_limits<unsigned long long>::type' (aka 'unsigned long long') to 'double' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-const-int-float-conversion]
11:45:40     if ((rhs > 0 && rhs < std::numeric_limits<uint64_t>::max()) ||
11:45:40                         ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11:45:40 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h:1310:1: error: reference to 'to_chars_result' is ambiguous
11:45:40 to_chars_result to_chars(char *first, char *last, const wide_integer<Bits, Signed> &value, int base) {
11:45:40 ^
11:45:40 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/mambaforge/envs/mantid-developer/bin/../include/c++/v1/__charconv/to_chars_result.h:24:34: note: candidate found by name lookup is 'std::__1::to_chars_result'
11:45:40 struct _LIBCPP_EXPORTED_FROM_ABI to_chars_result {
11:45:40                                  ^
11:45:40 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideInt.h:265:8: note: candidate found by name lookup is 'std::to_chars_result'
11:45:40 struct to_chars_result {
11:45:40        ^
11:45:40 1 warning and 2 errors generated.
```
```
11:45:43 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h:1312:19: error: no viable conversion from 'std::error_code' to 'std::errc'
11:45:43     return {last, std::make_error_code(std::errc::invalid_argument)};
11:45:43                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11:45:43 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h:1315:19: error: no viable conversion from 'std::error_code' to 'std::errc'
11:45:43     return {last, std::make_error_code(std::errc::invalid_argument)};
11:45:43                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11:45:43 /Users/builder/Jenkins/workspace/pull_requests-conda-osx/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h:1339:22: error: no viable conversion from 'std::error_code' to 'std::errc'
11:45:43     return {nullptr, std::make_error_code(std::errc::value_too_large)};
11:45:43                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
11:45:43 1 warning and 5 errors generated.
```

--- a/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideInt.h
+++ b/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideInt.h
@@ -22,10 +22,12 @@
  * without express or implied warranty.
  */
 
+#include <charconv>
 #include <climits> // CHAR_BIT
 #include <cmath>
 #include <cstdint>
 #include <iomanip>
+#include <ios>
 #include <limits>
 #include <ostream>
 #include <random>

--- a/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideInt.h
+++ b/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideInt.h
@@ -22,12 +22,10 @@
  * without express or implied warranty.
  */
 
-#include <charconv>
 #include <climits> // CHAR_BIT
 #include <cmath>
 #include <cstdint>
 #include <iomanip>
-#include <ios>
 #include <limits>
 #include <ostream>
 #include <random>
@@ -262,24 +260,6 @@ std::wostream &operator<<(std::wostream &out, const wide_integer<Bits, Signed> &
 template <size_t Bits, typename Signed> std::istream &operator>>(std::istream &in, wide_integer<Bits, Signed> &n);
 
 template <size_t Bits, typename Signed> std::wistream &operator>>(std::wistream &in, wide_integer<Bits, Signed> &n);
-
-//// Must be defined in another header
-struct to_chars_result {
-  char *ptr;
-  std::error_code ec;
-};
-
-struct from_chars_result {
-  const char *ptr;
-  std::error_code ec;
-};
-////
-
-template <size_t Bits, typename Signed>
-to_chars_result to_chars(char *first, char *last, const wide_integer<Bits, Signed> &value, int base = 10);
-
-template <size_t Bits, typename Signed>
-from_chars_result from_chars(const char *first, const char *last, wide_integer<Bits, Signed> &value, int base = 10);
 
 inline namespace literals {
 inline namespace wide_integer_literals {

--- a/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h
+++ b/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h
@@ -10,7 +10,9 @@
 
 #include <algorithm>
 #include <array>
+#include <charconv>
 #include <cstring>
+#include <ios>
 #include <iterator>
 #include <numeric>
 
@@ -180,8 +182,8 @@ template <size_t Bits, typename Signed> struct wide_integer<Bits, Signed>::_impl
   }
 
   constexpr static void wide_integer_from_bultin(wide_integer<Bits, Signed> &self, double rhs) noexcept {
-    if ((rhs > 0 && rhs < std::numeric_limits<uint64_t>::max()) ||
-        (rhs < 0 && rhs > std::numeric_limits<int64_t>::min())) {
+    if ((rhs > 0 && rhs < static_cast<double>(std::numeric_limits<uint64_t>::max())) ||
+        (rhs < 0 && rhs > static_cast<double>(std::numeric_limits<int64_t>::min()))) {
       self = to_Integral(rhs);
       return;
     }

--- a/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h
+++ b/Framework/DataObjects/inc/MantidDataObjects/MortonIndex/WideIntImpl.h
@@ -10,9 +10,7 @@
 
 #include <algorithm>
 #include <array>
-#include <charconv>
 #include <cstring>
-#include <ios>
 #include <iterator>
 #include <numeric>
 
@@ -1308,49 +1306,6 @@ template <size_t Bits, typename Signed> std::wistream &operator>>(std::wistream 
   return in;
 }
 
-template <size_t Bits, typename Signed>
-to_chars_result to_chars(char *first, char *last, const wide_integer<Bits, Signed> &value, int base) {
-  if (base < 2 || base > 36) {
-    return {last, std::make_error_code(std::errc::invalid_argument)};
-  }
-  if (first >= last) {
-    return {last, std::make_error_code(std::errc::invalid_argument)};
-  }
-
-  if (value == 0) {
-    *first = '0';
-    *(++first) = '\0';
-    return {++first, {}};
-  }
-
-  wide_integer<Bits, Signed> v = value;
-  if (v < 0) {
-    v = -v;
-    *(first++) = '-';
-  }
-
-  char *cur = last;
-
-  while (v != 0 && --cur >= first) {
-    static const char ALPHA[] = "0123456789abcdefghijklmnopqrstuvwxyz";
-    *cur = ALPHA[v % base];
-    v /= base;
-  }
-
-  if (v && cur + 1 == first) {
-    return {nullptr, std::make_error_code(std::errc::value_too_large)};
-  }
-
-  while (cur < last) {
-    *(first++) = *(cur++);
-  }
-  if (first < last) {
-    *first = '\0';
-  }
-
-  return {first, {}};
-}
-
 std::array<char, 256> inline genReverseAlpha() noexcept {
   static const char ALPHA[] = "0123456789abcdefghijklmnopqrstuvwxyz";
   std::array<char, 256> res;
@@ -1359,48 +1314,6 @@ std::array<char, 256> inline genReverseAlpha() noexcept {
     res[ALPHA[i]] = static_cast<char>(i);
   }
   return res;
-}
-
-template <size_t Bits, typename Signed>
-from_chars_result from_chars(const char *first, const char *last, wide_integer<Bits, Signed> &value, int base) {
-  if (base < 2 || base > 36) {
-    return {first, std::make_error_code(std::errc::invalid_argument)};
-  }
-  if (first >= last) {
-    return {first, std::make_error_code(std::errc::invalid_argument)};
-  }
-
-  bool is_negative = *first == '-';
-  if (is_negative) {
-    if (!is_same<Signed, signed>::value) {
-      return {first, std::make_error_code(std::errc::result_out_of_range)};
-    }
-    if (++first >= last) {
-      return {first, std::make_error_code(std::errc::invalid_argument)};
-    }
-  }
-
-  wide_integer<Bits, Signed> v = 0;
-  const char *cur = first;
-
-  do {
-    static const std::array<char, 256> ALPHA = genReverseAlpha();
-    char cv = ALPHA[*cur];
-    if (cv >= base || cv == -1) {
-      if (cur == first) {
-        return {cur, std::make_error_code(std::errc::result_out_of_range)};
-      } else {
-        value = v;
-        return {cur, {}};
-      }
-    }
-
-    v *= base;
-    v += cv;
-  } while (++cur < last);
-
-  value = is_negative ? -v : v;
-  return {cur, {}};
 }
 
 constexpr int128_t operator"" _cppi128(const char *n) { return int128_t::_impl::from_str(n); }

--- a/Framework/DataObjects/test/MDLeanEventTest.h
+++ b/Framework/DataObjects/test/MDLeanEventTest.h
@@ -7,7 +7,6 @@
 #pragma once
 
 #include "MantidDataObjects/MDLeanEvent.h"
-#include "MantidDataObjects/MortonIndex/WideInt.h"
 #include "MantidKernel/System.h"
 #include "MantidKernel/Timer.h"
 #include <cxxtest/TestSuite.h>

--- a/mantid-developer-osx.yml
+++ b/mantid-developer-osx.yml
@@ -51,7 +51,7 @@ dependencies:
   - black  # may be out of sync with pre-commit
   - cppcheck==2.13.1
   - pre-commit>=2.12.0
-  - libcxx<17 #Had to pin as there is an issue with 'to_chars_result'
+  - libcxx
   - joblib
   - orsopy==1.2.0 # Fix the version to avoid updates being pulled in automatically, which might change the Reflectometry ORSO file content or layout and cause tests to fail.
 

--- a/qt/widgets/common/test/AlgorithmRunnerTest.h
+++ b/qt/widgets/common/test/AlgorithmRunnerTest.h
@@ -29,6 +29,8 @@ IConfiguredAlgorithm_sptr createConfiguredAlgorithm() {
   return configuredAlg;
 }
 
+} // namespace
+
 GNU_DIAG_OFF_SUGGEST_OVERRIDE
 
 class MockAlgorithmRunnerSubscriber : public IAlgorithmRunnerSubscriber {
@@ -48,8 +50,6 @@ MATCHER(CheckAlgorithmNull, "Check the algorithm is a nullptr") { return !arg; }
 MATCHER_P(CheckQueueSize, size, "Check the size of a std::deque") { return arg.size() == size; }
 
 GNU_DIAG_ON_SUGGEST_OVERRIDE
-
-} // namespace
 
 class AlgorithmRunnerTest : public CxxTest::TestSuite {
 public:

